src/b64.ml

```diff
@@ -16,77 +19,153 @@
  *
  *)
 
-let default_alphabet =
-  "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
-
-let uri_safe_alphabet =
-  "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_"
-
-let padding = '='
-
-let of_char ?(alphabet = default_alphabet) x =
-  if x = padding then 0 else String.index alphabet x
-
-let to_char ?(alphabet = default_alphabet) x = alphabet.[x]
-
-let decode ?alphabet input =
-  let length = String.length input in
-  let input =
-    if length mod 4 = 0 then input
-    else input ^ String.make (4 - (length mod 4)) padding
-  in
-  let length = String.length input in
-  let words = length / 4 in
-  let padding =
-    match length with
-    | 0 -> 0
-    | _ when input.[length - 2] = padding -> 2
-    | _ when input.[length - 1] = padding -> 1
-    | _ -> 0
-  in
-  let output = Bytes.make ((words * 3) - padding) '\000' in
-  for i = 0 to words - 1 do
-    let a = of_char ?alphabet input.[(4 * i) + 0]
-    and b = of_char ?alphabet input.[(4 * i) + 1]
-    and c = of_char ?alphabet input.[(4 * i) + 2]
-    and d = of_char ?alphabet input.[(4 * i) + 3] in
-    let n = (a lsl 18) lor (b lsl 12) lor (c lsl 6) lor d in
-    let x = (n lsr 16) land 255
-    and y = (n lsr 8) land 255
-    and z = n land 255 in
-    Bytes.set output ((3 * i) + 0) (char_of_int x) ;
-    if i <> words - 1 || padding < 2 then
-      Bytes.set output ((3 * i) + 1) (char_of_int y) ;
-    if i <> words - 1 || padding < 1 then
-      Bytes.set output ((3 * i) + 2) (char_of_int z)
-  done ;
-  Bytes.unsafe_to_string output
+type alphabet =
+  { emap : int array
+  ; dmap : int array }
+
+let (//) x y =
+  if y < 1 then raise Division_by_zero ;
+  if x > 0 then 1 + ((x - 1) / y) else 0
+[@@inline]
+
+let unsafe_get_uint8 t off = Char.code (String.unsafe_get t off)
+let unsafe_set_uint8 t off v = Bytes.unsafe_set t off (Char.chr v)
+
+external unsafe_set_uint16 : bytes -> int -> int -> unit = "%caml_string_set16u" [@@noalloc]
+external swap16 : int -> int = "%bswap16" [@@noalloc]
+
+let none = (-1)
+
+let padding_exists alphabet = String.contains alphabet '='
+
+let make_alphabet alphabet =
+  if String.length alphabet <> 64 then invalid_arg "Length of alphabet must be 64" ;
+  if padding_exists alphabet then invalid_arg "Alphabet can not contain padding character" ;
+  let emap = Array.init (String.length alphabet) (unsafe_get_uint8 alphabet)  in
+  let dmap = Array.make 256 none in
+  String.iteri (fun idx chr -> Array.unsafe_set dmap (Char.code chr) idx) alphabet ;
```

Comment:
Making an alphabet is really not performance sensitive, so I'd prefer to remove these unsafe get/sets here if possible...